test/sql/test_cte.py

```diff
@@ -1377,3 +1379,306 @@ def test_no_alias_construct(self):
+        select_1_cte = select([literal(1).label("inner_cte")]).cte(
+            "nesting_1", nesting=True
+        )
+        select_2_cte = select([literal(2).label("inner_cte")]).cte(
+            "nesting_2", nesting=True
+        )
+
+        nesting_cte = (
+            select([select_1_cte]).union(select([select_2_cte])).subquery()
+        )
+
+        stmt = select(
+            [select([nesting_cte.c.inner_cte.label("outer_cte")]).cte("cte")]
+        )
+
+        self.assert_compile(
+            stmt,
+            "WITH cte AS ("
+            "SELECT anon_1.inner_cte AS outer_cte FROM ("
+            "WITH nesting_1 AS (SELECT %(param_1)s AS inner_cte)"
+            ", nesting_2 AS (SELECT %(param_2)s AS inner_cte)"
+            " SELECT nesting_1.inner_cte AS inner_cte FROM nesting_1"
+            " UNION"
+            " SELECT nesting_2.inner_cte AS inner_cte FROM nesting_2"
+            ") AS anon_1"
+            ") SELECT cte.outer_cte FROM cte",
+        )
+
+    def test_nesting_cte_in_recursive_cte(self):
+        nesting_cte = select([literal(1).label("inner_cte")]).cte(
+            "nesting", nesting=True
+        )
+        stmt = select(
+            [
+                select([nesting_cte.c.inner_cte.label("outer_cte")]).cte(
+                    "cte", recursive=True
+                )
+            ]
+        )
+
+        self.assert_compile(
+            stmt,
+            "WITH RECURSIVE cte(outer_cte) AS (WITH nesting AS "
+            "(SELECT %(param_1)s AS inner_cte) "
+            "SELECT nesting.inner_cte AS outer_cte FROM nesting) "
+            "SELECT cte.outer_cte FROM cte",
+        )
+
+    def test_recursive_nesting_cte_in_cte(self):
+        nesting_cte = select([literal(1).label("inner_cte")]).cte(
+            "nesting", nesting=True, recursive=True
+        )
+        stmt = select(
+            [select([nesting_cte.c.inner_cte.label("outer_cte")]).cte("cte")]
+        )
+
+        self.assert_compile(
+            stmt,
+            "WITH cte AS (WITH RECURSIVE nesting(inner_cte) AS "
+            "(SELECT %(param_1)s AS inner_cte) "
+            "SELECT nesting.inner_cte AS outer_cte FROM nesting) "
+            "SELECT cte.outer_cte FROM cte",
+        )
+
+    def test_recursive_nesting_cte_in_recursive_cte(self):
+        nesting_cte = select([literal(1).label("inner_cte")]).cte(
+            "nesting", nesting=True, recursive=True
+        )
+        stmt = select(
+            [
+                select([nesting_cte.c.inner_cte.label("outer_cte")]).cte(
+                    "cte", recursive=True
+                )
+            ]
+        )
+
+        self.assert_compile(
+            stmt,
+            "WITH RECURSIVE cte(outer_cte) AS (WITH RECURSIVE nesting(inner_cte) "
+            "AS (SELECT %(param_1)s AS inner_cte) "
+            "SELECT nesting.inner_cte AS outer_cte FROM nesting) "
+            "SELECT cte.outer_cte FROM cte",
+        )
+
+    @pytest.mark.parametrize(
```

Comment:
I dont think we should raise within the compiler for this condition, so IMO we don't need this test.

Also, we dont use `pytest.mark.parametrize` directly, we use our own abstraction `@testing.combinations`, see https://github.com/sqlalchemy/sqlalchemy/blob/master/test/sql/test_cte.py#L1032 in this file for example

Reply:
Just for curiosity why do you use `@testing.combinations` instead of `pytest.mark.parametrize` ?

Reply:
The idea is to avoid an hard dependency on pytest.
If I'm not mistaken sa has already outlived two other testing framework before using pytest.

Also Mike likes doing stuff it's own way 😅

Reply:
yes originally the idea was, the test suite ran both on nose and pytest.  this is when pytest was new.      once the enormous job of migrating off of nose into a generic abstraction layer was done, we weren't going to have to do that again, so we kept the abstraction layer.

later, we started using way more pytest features.  but we built those into the abstraction layer also.

this continues to be a good thing, because:

1. it limits us to using specific pytest patterns that we have decided to adopt, rather than using the whole kitchen sink and having different approaches in different tests
2. it completely saves our bacon when we hit issues such as the older pytest 4.x series that is the newest version that runs on python 2 has some serious performance issues with certain kinds of fixtures; the abstraction layer makes sure we don't use those features
3. keeps us honest and organized about how more elaborate pytest features are used

so shorter bullets:

1. pytest is vast, our abstraction layer helps with the issue of tests from using dozens of different patterns everywhere (which they still do, but not as much as they would without the abstraction)
2. gives us a little bit of insurance for migrating to new versions of pytest and/or a new test framework if pytest goes south someday